preprocessing/data_science_help_functions.py

```diff
@@ -189,3 +196,140 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
+    adversarial_validation_result, clf = get_adv_validation_score(df_joined, y)
+
+    # Process result:
+    if adversarial_validation_result > threshold:
+        print(
+            f"WARNING! There is significant difference between {list(dataframe_dict.keys())[0]} and {list(dataframe_dict.keys())[0]}\n"
+            f"datasets in terms of feature distribution. Validation score: {adversarial_validation_result}, threshold: {threshold}")
+        print(f"Top features are: {xgb_important_features(clf)}")
+    else:
+        print(
+            f"There is no significant difference between {list(dataframe_dict.keys())[0]} and {list(dataframe_dict.keys())[0]}\n"
+            f"datasets in terms of feature distribution. Validation score: {adversarial_validation_result}, threshold: {threshold}")
+    return adversarial_validation_result
+
+
+def get_adv_validation_score(df_joined: pd.DataFrame,
+                             y: pd.Series) -> Tuple[float, xgb.sklearn.XGBClassifier]:
+    """ Calculate advisarial validation score based on dataframes and XGBClassifier
+
+    :param DataFrame df_joined: Feature dataframe
+    :param Series y: Target series
+
+    :return:
+            | clf: Trained model
+            | mean of KFold validation results (ROC-AUC scores)
+    """
+
+    skf = StratifiedKFold(n_splits=3, shuffle=True, random_state=44)
+    xgb_params = {
+        'learning_rate': 0.1, 'max_depth': 4, 'subsample': 0.9,
+        'colsample_bytree': 0.9, 'objective': 'binary:logistic',
+        'silent': 1, 'n_estimators': 50, 'gamma': 1,
+        'min_child_weight': 4
+    }
+    clf = xgb.XGBClassifier(**xgb_params, seed=10)
+    results = []
+    logger.info('Adversarial validation checking:')
+    for fold, (train_index, test_index) in enumerate(skf.split(df_joined, y)):
+        fold_xtrain, fold_xval = df_joined.iloc[train_index], df_joined.iloc[test_index]
+        fold_ytrain, fold_yval = y.iloc[train_index], y.iloc[test_index]
+        clf.fit(fold_xtrain, fold_ytrain, eval_set=[(fold_xval, fold_yval)],
+                eval_metric='logloss', verbose=False, early_stopping_rounds=10)
+        fold_ypred = clf.predict_proba(fold_xval)[:, 1]
+        fold_score = roc_auc_score(fold_yval, fold_ypred)
+        results.append(fold_score)
+        logger.info(f"Fold: {fold + 1} shape: {fold_xtrain.shape} score: {fold_score}")
+
+    return round(np.mean(results), 2), clf
+
+
+def xgb_important_features(xgb: xgb.sklearn.XGBClassifier,
+                           top_features: int = 5) -> str:
+    """ Get top of the most important features from a trained model
+
+    :param XGBClassifier xgb: A trained model
+    :param int top_features: Max length of features to send back
+
+    :return:
+            | A string with a list of the most important features plus their importance
+    """
+
+    # get features
+    feat_imp = xgb.get_booster().get_score(importance_type='gain')
+
+    # round importances
+    for dict_key in feat_imp:
```

Comment:
I would suggest extracting it in a new small function

Reply:
done